### ISSUE TITLE
Slice error

### ISSUE THREAD

**Issue by c1au6i0** [EGA-archive/ega-download-client#120]
**Slice error** 

Error arises when using multiple connections to download single files:
```pyega3 -c 30 -cf  ega_credentials.json fetch EGAF00001753775 --saveto EGAF00001753775```

Error:
```
[2021-08-08 19:09:01 -0400] Slice error: received=266796603, requested=472370309, file='EGAF00001753775-from-472370309-len-472370309.slice'
Traceback (most recent call last):
  File "/home/clz4002/miniconda3/lib/python3.8/site-packages/pyega3/pyega3.py", line 494, in download_file_retry
    download_file(token, file_id, file_size, check_sum, num_connections, key, output_file)
  File "/home/clz4002/miniconda3/lib/python3.8/site-packages/pyega3/pyega3.py", line 422, in download_file
    for part_file_name in executor.map(download_file_slice_, params):
  File "/home/clz4002/miniconda3/lib/python3.8/concurrent/futures/_base.py", line 611, in result_iterator
    yield fs.pop().result()
  File "/home/clz4002/miniconda3/lib/python3.8/concurrent/futures/_base.py", line 439, in result
    return self.__get_result()
  File "/home/clz4002/miniconda3/lib/python3.8/concurrent/futures/_base.py", line 388, in __get_result
    raise self._exception
  File "/home/clz4002/miniconda3/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/clz4002/miniconda3/lib/python3.8/site-packages/pyega3/pyega3.py", line 296, in download_file_slice_
    return download_file_slice(*args)
  File "/home/clz4002/miniconda3/lib/python3.8/site-packages/pyega3/pyega3.py", line 290, in download_file_slice
    raise Exception("Slice error: received={}, requested={}, file='{}'".format(total_received, length, file_name))
Exception: Slice error: received=266796603, requested=472370309, file='EGAF00001753775-from-472370309-len-472370309.slice'
Traceback (most recent call last):
  File "/home/clz4002/miniconda3/bin/pyega3", line 10, in <module>
    sys.exit(main())
  File "/home/clz4002/miniconda3/lib/python3.8/site-packages/pyega3/pyega3.py", line 671, in main
    download_file_retry(credentials, args.identifier, display_file_name, file_name, file_size, check_sum, args.connections, key,
  File "/home/clz4002/miniconda3/lib/python3.8/site-packages/pyega3/pyega3.py", line 502, in download_file_retry
    raise e
  File "/home/clz4002/miniconda3/lib/python3.8/site-packages/pyega3/pyega3.py", line 494, in download_file_retry
    download_file(token, file_id, file_size, check_sum, num_connections, key, output_file)
  File "/home/clz4002/miniconda3/lib/python3.8/site-packages/pyega3/pyega3.py", line 422, in download_file
    for part_file_name in executor.map(download_file_slice_, params):
  File "/home/clz4002/miniconda3/lib/python3.8/concurrent/futures/_base.py", line 611, in result_iterator
    yield fs.pop().result()
  File "/home/clz4002/miniconda3/lib/python3.8/concurrent/futures/_base.py", line 439, in result
    return self.__get_result()
  File "/home/clz4002/miniconda3/lib/python3.8/concurrent/futures/_base.py", line 388, in __get_result
    raise self._exception
  File "/home/clz4002/miniconda3/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/clz4002/miniconda3/lib/python3.8/site-packages/pyega3/pyega3.py", line 296, in download_file_slice_
    return download_file_slice(*args)
  File "/home/clz4002/miniconda3/lib/python3.8/site-packages/pyega3/pyega3.py", line 290, in download_file_slice
    raise Exception("Slice error: received={}, requested={}, file='{}'".format(total_received, length, file_name))
Exception: Slice error: received=266796603, requested=472370309, file='EGAF00001753775-from-472370309-len-472370309.slice'
```
After 5 attempts download fails.

I have not noticed this particular error (same lines) in previously reported bugs

-  Linux 3.10.0-1127.19.1.el7.x86_64
 - Python version:  Python 3.8.5
 - PyEGA3 version: 3.4.1
 - Installation tested

**Comment by CocoMlle** [comment]
Hi, i have the exact same error, does any solution occured? 

Python version : 3.6.2
PyEGA3 version: 3.4.1
Installation tested with success

[2021-08-23 22:10:26 +0200] Download starting [using 1 connection(s)]...
[2021-08-24 05:02:41 +0200] Slice error: received=22415664216, requested=85122315797, file='./EGAF00001723716/NA15850.bam-from-0-len-85122315797.slice'
Traceback (most recent call last):
  File "/work/gad/shared/bin/miniconda3/lib/python3.6/site-packages/pyega3/pyega3.py", line 494, in download_file_retry
    download_file(token, file_id, file_size, check_sum, num_connections, key, output_file)
  File "/work/gad/shared/bin/miniconda3/lib/python3.6/site-packages/pyega3/pyega3.py", line 422, in download_file
    for part_file_name in executor.map(download_file_slice_, params):
  File "/work/gad/shared/bin/miniconda3/lib/python3.6/concurrent/futures/_base.py", line 556, in result_iterator
    yield future.result()
  File "/work/gad/shared/bin/miniconda3/lib/python3.6/concurrent/futures/_base.py", line 405, in result
    return self.__get_result()
  File "/work/gad/shared/bin/miniconda3/lib/python3.6/concurrent/futures/_base.py", line 357, in __get_result
    raise self._exception
  File "/work/gad/shared/bin/miniconda3/lib/python3.6/concurrent/futures/thread.py", line 55, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/work/gad/shared/bin/miniconda3/lib/python3.6/site-packages/pyega3/pyega3.py", line 296, in download_file_slice_
    return download_file_slice(*args)
  File "/work/gad/shared/bin/miniconda3/lib/python3.6/site-packages/pyega3/pyega3.py", line 290, in download_file_slice
    raise Exception("Slice error: received={}, requested={}, file='{}'".format(total_received, length, file_name))
Exception: Slice error: received=22415664216, requested=85122315797, file='./EGAF00001723716/NA15850.bam-from-0-len-85122315797.slice'
[2021-08-24 05:03:41 +0200] retry attempt 3
i tried many times but i still have the same issue and i really need these data

**Comment by BinghuaLi** [comment]
[2022-07-16 13:35:11 +0800] Slice error: received=100663296, requested=104857600, file='/home/lbh/Projects/EGA/TplusA/Rawdata/RNAseq/EGAF00005481342/.tmp_download/EGAF00005481342-from-629145600-len-104857600.slice.tmp'
Traceback (most recent call last):
  File "/home/lbh/miniconda3/envs/RNA-seq/lib/python3.6/site-packages/pyega3/libs/data_file.py", line 304, in download_file_retry
    self.download_file(output_file, num_connections, max_slice_size)
  File "/home/lbh/miniconda3/envs/RNA-seq/lib/python3.6/site-packages/pyega3/libs/data_file.py", line 143, in download_file
    for part_file_name in executor.map(self.download_file_slice_, params):
  File "/home/lbh/miniconda3/envs/RNA-seq/lib/python3.6/concurrent/futures/_base.py", line 586, in result_iterator
    yield fs.pop().result()
  File "/home/lbh/miniconda3/envs/RNA-seq/lib/python3.6/concurrent/futures/_base.py", line 432, in result
    return self.__get_result()
  File "/home/lbh/miniconda3/envs/RNA-seq/lib/python3.6/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "/home/lbh/miniconda3/envs/RNA-seq/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/lbh/miniconda3/envs/RNA-seq/lib/python3.6/site-packages/pyega3/libs/data_file.py", line 173, in download_file_slice_
    return self.download_file_slice(*args)
  File "/home/lbh/miniconda3/envs/RNA-seq/lib/python3.6/site-packages/pyega3/libs/data_file.py", line 216, in download_file_slice
    raise Exception(f"Slice error: received={total_received}, requested={length}, file='{file_name}'")
Exception: Slice error: received=100663296, requested=104857600, file='/home/lbh/Projects/EGA/TplusA/Rawdata/RNAseq/EGAF00005481342/.tmp_download/EGAF00005481342-from-629145600-len-104857600.slice.tmp'
[2022-07-16 13:36:11 +0800] retry attempt 5


simmilar error.

**Comment by xeeloij** [comment]
Same error here. Download terminated after 5 attempt... BTW, no reply from the EGA Helpdesk. Can anyone help me??

**Comment by c1au6i0** [comment]
They released a new client that might solve some problems. Back in the days I used 
to  "downloaded each file of the dataset in parallel/at the same time (but with only 1 connection). Some will fail after the 5th attempt, but other will succeed. You repeat the operation few times and eventually in few days you will be able to download all the files".  Yes, i had another issue and I think tha

**Comment by Prathyusha-konda** [comment]
Same slice error and trouble with downloading, I only installed pyega3 last week so this should be a recent update.

[2023-01-20 14:08:54 -0500] Retrying (Retry(total=19, connect=False, read=9, redirect=None, status=10)) after connection broken by 'ProtocolError('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))': /v2/files/EGAF00001883151?destinationFormat=plain
[2023-01-20 14:14:55 -0500] Slice error: received=55881985, requested=104857600, file='/mnt/storage/labs/sviswanathan/data/prats/simon/EGAF00001883141/.tmp_download/EGAF00001883141-from-104857600-len-104857600.slice.tmp'
Traceback (most recent call last):
  File "/home/pk576/miniconda3/lib/python3.9/site-packages/pyega3/libs/data_file.py", line 326, in download_file_retry
    self.download_file(output_file, num_connections, max_slice_size)
  File "/home/pk576/miniconda3/lib/python3.9/site-packages/pyega3/libs/data_file.py", line 159, in download_file
    for part_file_name in executor.map(self.download_file_slice_, params):
  File "/home/pk576/miniconda3/lib/python3.9/concurrent/futures/_base.py", line 609, in result_iterator
    yield fs.pop().result()
  File "/home/pk576/miniconda3/lib/python3.9/concurrent/futures/_base.py", line 446, in result
    return self.__get_result()
  File "/home/pk576/miniconda3/lib/python3.9/concurrent/futures/_base.py", line 391, in __get_result
    raise self._exception
  File "/home/pk576/miniconda3/lib/python3.9/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/pk576/miniconda3/lib/python3.9/site-packages/pyega3/libs/data_file.py", line 189, in download_file_slice_
    return self.download_file_slice(*args)
  File "/home/pk576/miniconda3/lib/python3.9/site-packages/pyega3/libs/data_file.py", line 234, in download_file_slice
    raise Exception(f"Slice error: received={total_received}, requested={length}, file='{file_name}'")
Exception: Slice error: received=55881985, requested=104857600, file='/mnt/storage/labs/sviswanathan/data/prats/simon/EGAF00001883141/.tmp_download/EGAF00001883141-from-104857600-len-104857600.slice.tmp'
[2023-01-20 14:14:56 -0500] Slice error: received=48428142, requested=104857600, file='/mnt/storage/labs/sviswanathan/data/prats/simon/EGAF00001883130/.tmp_download/EGAF00001883130-from-0-len-104857600.slice.tmp'
Traceback (most recent call last):
  File "/home/pk576/miniconda3/lib/python3.9/site-packages/pyega3/libs/data_file.py", line 326, in download_file_retry
    self.download_file(output_file, num_connections, max_slice_size)
  File "/home/pk576/miniconda3/lib/python3.9/site-packages/pyega3/libs/data_file.py", line 159, in download_file
    for part_file_name in executor.map(self.download_file_slice_, params):
  File "/home/pk576/miniconda3/lib/python3.9/concurrent/futures/_base.py", line 609, in result_iterator
    yield fs.pop().result()
  File "/home/pk576/miniconda3/lib/python3.9/concurrent/futures/_base.py", line 446, in result
    return self.__get_result()
  File "/home/pk576/miniconda3/lib/python3.9/concurrent/futures/_base.py", line 391, in __get_result
    raise self._exception
  File "/home/pk576/miniconda3/lib/python3.9/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/pk576/miniconda3/lib/python3.9/site-packages/pyega3/libs/data_file.py", line 189, in download_file_slice_
    return self.download_file_slice(*args)
  File "/home/pk576/miniconda3/lib/python3.9/site-packages/pyega3/libs/data_file.py", line 234, in download_file_slice
    raise Exception(f"Slice error: received={total_received}, requested={length}, file='{file_name}'")

**Comment by zengxi-hada** [comment]
I also experienced the same problems for both 1 and 5 connections. Any solutions?
   
 Slice error: received=20633125, requested=104857600, 
     file='/data/EGAF00001197014/
     17 Traceback (most recent call last):
     18   File "/home/xiz978/anaconda3/lib/python3.9/site-packages/pyega3/libs/data_file.py", line 326, in download_file_retry
     19     self.download_file(output_file, num_connections, max_slice_size)
     20   File "/home/xiz978/anaconda3/lib/python3.9/site-packages/pyega3/libs/data_file.py", line 159, in download_file
     21     for part_file_name in executor.map(self.download_file_slice_, params):
     22   File "/home/xiz978/anaconda3/lib/python3.9/concurrent/futures/_base.py", line 609, in result_iterator
     23     yield fs.pop().result()
     24   File "/home/xiz978/anaconda3/lib/python3.9/concurrent/futures/_base.py", line 446, in result
     25     return self.__get_result()
     26   File "/home/xiz978/anaconda3/lib/python3.9/concurrent/futures/_base.py", line 391, in __get_result
     27     raise self._exception
     28   File "/home/xiz978/anaconda3/lib/python3.9/concurrent/futures/thread.py", line 58, in run
     29     result = self.fn(*self.args, **self.kwargs)
     30   File "/home/xiz978/anaconda3/lib/python3.9/site-packages/pyega3/libs/data_file.py", line 189, in download_file_slice_
     31     return self.download_file_slice(*args)
     32   File "/home/xiz978/anaconda3/lib/python3.9/site-packages/pyega3/libs/data_file.py", line 234, in download_file_slice
     33     raise Exception(f"Slice error: received={total_received}, requested={length}, file='{file_name}'")
     34 Exception: Slice error: received=20633125, requested=104857600, file='/n/scratch3/users/x/xiz978/L1_mediated_SV/data_from_ICGC/cancer_bulk_WGS/data/EGAF00001197014/.tmp_download/EGA
     35 [2023-02-11 20:00:30 -0500] retry attempt 1

**Comment by stefanpeidli** [comment]
Same error here. Downloading single files instead of datasets and using `-c 1` also fails. 

Actually, downloading literally any file with pyEGA yields this Slice error approx. every 5mins and eventually (after 5 retries) stops the download. The largest I could get into a download was around 130MB.

**Comment by suhartobanerjee** [comment]
Having the same issues. Additionally, the server keeps disconnecting:

`[2023-02-24 17:28:25 +0100] Retrying (Retry(total=19, connect=False, read=9, redirect=None, status=10)) after connection broken by 'ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))': /v2/files/EGAF00005686531?destinationFormat=plain`

I am trying with 10mb chunks, max I got is 130MB like @stefanpeidli mentioned!

**Comment by yanchunzhang** [comment]
> Hi @c1au6i0 , @CocoMlle , @benbfly - Thank you all for reporting that you are experiencing this issue, and apologies that this is causing delays in retrieving the files you are after. A few suggestions to help you acquire the files:
> 
> * As @c1au6i0 mentioned, using only one connection (`-c 1`) for smaller files. There are various retries built into pyega3 and the underlying API, and we are currently integrating even more retries so that in the future you won't have to retry manually on your side.
> * Select genomic regions of interest (if applicable to your needs) for larger files. This [feature](https://github.com/EGA-archive/ega-download-client#using-pyega3-for-fetching-a-genomic-range) (`pyega3 fetch`) is available for VCF and BAM files which have indexes archived, and greatly reduces the size of the data downloaded via pyega3.
> * Help us test our improved API/pyega3 service when it is deployed in a few months :)
> 
> Because @c1au6i0 is eventually able to get the file downloaded, I suspect the issue is with our underlying API, and not pyega3 itself. We are in the final stages of testing a new version of our API, which should provide a more stable data download service. Thank you for using pyega3 and providing this valuable feedback!

I don't think the one connection settting will solve the problem. What should I do if I want to download very big files?
Yanchun

**Comment by bertagayete** [comment]
I am also facing this problem and the suggestions of @malloryfreeberg do not solve the issue.

Command:
`pyega3 -c 1 -cf /media/sequentia/sdc1/visitor12/CREDENTIALS_FILE.JSON fetch EGAF00002444145 --output-dir $PWD`
 
Error:
```
[2023-05-15 12:37:22 +0000] 
[2023-05-15 12:37:22 +0000] pyEGA3 - EGA python client version 5.0.2 (https://github.com/EGA-archive/ega-download-client) 
[2023-05-15 12:37:22 +0000] Parts of this software are derived from pyEGA (https://github.com/blachlylab/pyega) by James Blachly 
[2023-05-15 12:37:22 +0000] Python version : 3.9.15 
[2023-05-15 12:37:22 +0000] OS version : Linux #78-Ubuntu SMP Tue Apr 18 09:00:29 UTC 2023 
[2023-05-15 12:37:22 +0000] Server URL: https://ega.ebi.ac.uk:8443/v2 
[2023-05-15 12:37:22 +0000] Session-Id: 931592480 
[2023-05-15 12:37:23 +0000] 
[2023-05-15 12:37:23 +0000] Authentication success for user 'XXXXXXXXXXX' [2023-05-15 12:37:30 +0000] File Id: 'EGAF00002444145'(6354045145 bytes). 
[2023-05-15 12:37:30 +0000] Total space : 146.59 GiB 
[2023-05-15 12:37:30 +0000] Used space : 47.96 GiB 
[2023-05-15 12:37:30 +0000] Free space : 91.11 GiB 
[2023-05-15 12:37:30 +0000] Download starting [using 5 connection(s), file size 6354045129 and chunk length 104857600]...   
0%|                                                                                                                                    
                                                                                                       | 0.00/6.35G [00:00<?, ?B/s][2023-05-15 12:42:30 +0000] 
Retrying (Retry(total=19, connect=False, read=9, redirect=None, status=10)) after connection broken by 'ProtocolError('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))': /v2/files/EGAF00002444145?destinationFormat=plain 
[2023-05-15 12:44:27 +0000] Retrying (Retry(total=19, connect=False, read=9, redirect=None, status=10)) after connection broken by 'ProtocolError('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))': /v2/files/EGAF00002444145?destinationFormat=plain 
[2023-05-15 12:49:30 +0000] Retrying (Retry(total=18, connect=False, read=8, redirect=None, status=10)) after connection broken by 'ProtocolError('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))': /v2/files/EGAF00002444145?destinationFormat=plain   
0%|▏                                                                                                                      
                                                                                                                    | 9.83M/6.35G [18:40<70:29:41, 25.0kB/s] 
[2023-05-15 12:57:01 +0000] Slice error: received=0, requested=104857600, file='/media/sequentia/sdc1/visitor12/EGAF00002444145/.tmp_download/EGAF00002444145-from-0-len-104857600.slice.tmp' Traceback (most recent call last):   
      File "/home/asquitieri/.local/lib/python3.9/site-packages/pyega3/libs/data_file.py", line 323, in download_file_retry     self.download_file(output_file, num_connections, max_slice_size)   
      File "/home/asquitieri/.local/lib/python3.9/site-packages/pyega3/libs/data_file.py", line 159, in download_file     for part_file_name in executor.map(self.download_file_slice_, params):   
      File "/miniconda3/lib/python3.9/concurrent/futures/_base.py", line 609, in result_iterator     yield fs.pop().result()   
      File "/miniconda3/lib/python3.9/concurrent/futures/_base.py", line 446, in result     return self.__get_result()   File "/miniconda3/lib/python3.9/concurrent/futures/_base.py", line 391, in __get_result     raise self._exception   
      File "/miniconda3/lib/python3.9/concurrent/futures/thread.py", line 58, in run     result = self.fn(*self.args, **self.kwargs)   
      File "/home/asquitieri/.local/lib/python3.9/site-packages/pyega3/libs/data_file.py", line 189, in download_file_slice_     return self.download_file_slice(*args)   
      File "/home/asquitieri/.local/lib/python3.9/site-packages/pyega3/libs/data_file.py", line 231, in download_file_slice     raise Exception(f"Slice error: received={total_received}, requested={length}, file='{file_name}'") 
Exception: Slice error: received=0, requested=104857600, file='/media/sequentia/sdc1/visitor12/EGAF00002444145/.tmp_download/EGAF00002444145-from-0-len-104857600.slice.tmp' ```


Does anyone knows how to solve this issue?

**Comment by raagagrawal** [comment]
Was there an update on this issue. I am also seeing the same thing in many runs.

**Comment by sufyazi** [comment]
Any updates on this? I am using 5.0.2 version, which I believe is the latest and I keep getting this error, and this is very cumbersome when trying to download a large dataset that I was granted access to. Any solutions or at least has this been addressed by any of the devs for the 5.0.2 version?

**Comment by zengxi-hada** [comment]
It seems that they have updated it at https://github.com/EGA-archive/ega-download-client 
I test it and it seems that now it can work.

**Comment by raagagrawal** [comment]
I am using the 5.0.2 version and it is not working.

**Comment by Storyboardslee** [comment]
^Same here! I am also having issues with the slice error, and I am using 5.0.2. It has been a long time and after multiple retry attempt, the error is still occurring. Any suggestions?

**Comment by jeantristanb** [comment]
I have same issue, any news? there is other way to download data?

**Comment by CsabaHalmagyi** [comment]
There is a fix for slice errors already in production, please try downloading the files again. Should you encounter any issues, please contact our helpdesk team directly at [helpdesk@ega-archive.org](mailto:helpdesk@ega-archive.org), they can offer alternative ways of accessing the data.

**Comment by gregory-numenos** [comment]
@CsabaHalmagyi Same for me. Didn't get any response from the helpdesk. I need one specific file, that i've tried to download innumerous number of times with different number of connections and retries. Is there any other way to download it yet?
